tracing: allow background tracing via trace.mode cluster

tbg opened the pull request:
When `trace.mode` is set to `background`, our tracer

1. never creates noop spans
2. always returns a recording from `GetRecording()`

The recording in 2) is usually only minimally populated as
the spans are not automatically put in verbose mode.

This is introduced as a cluster setting now for ease of perf optimizing,
which will be necessary if we want to make background tracing the
default, as well to introducing structured metadata (#55584) and
testing. The hope is that we can remove it before the release;
it's not marked as public.

Relates to #55584.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57292)
<!-- Reviewable:end -->

tbg committed:
tracing: allow background tracing via trace.mode cluster setting

When `trace.mode` is set to `background`, our tracer

1. never creates noop spans
2. always returns a recording from `GetRecording()`

The recording in 2) is usually only minimally populated as
the spans are not automatically put in verbose mode.

This is introduced as a cluster setting now for ease of perf optimizing,
which will be necessary if we want to make background tracing the
default, as well to introducing structured metadata (#55584) and
testing. The hope is that we can remove it before the release;
it's not marked as public.

Here's the setting description:

    configures the CockroachDB-internal tracing subsystem.

    If set to 'background', trace spans will be created for all operations, but
    these trace spans will only be recording sparse structured information,
    unless an operation explicitly requests verbose recording. This is
    optimized for low overhead, and powers fine-grained statistics and alerts.

    If set to 'legacy', trace spans will not be created (unless an
    auxiliary tracer such as Lightstep or Zipkin, is configured, or an
    internal code path explicitly requests a trace to be created) but
    when they are, they record verbose information. This has two effects:
    the observability of the cluster may be degraded (as some trace spans
    are elided) and where trace spans are created, they may consume large
    amounts of memory. This mode should not be used with auxiliary tracing
    sinks as that leads to expensive trace spans being created throughout.

    Note that regardless of this setting, configuring an auxiliary
    trace sink will cause verbose traces to be created for all
    operations, which may lead to high memory consumption. It is not
    currently possible to send non-verbose traces to auxiliary sinks.

Relates to #55584.

Release note: None

tbg commented:
Current top10 in

```
go tool pprof -sample_index alloc_objects -diff_base false.prof true.prof
```

is

```
(pprof) top10
Showing nodes accounting for 89472, 1.22% of 7310530 total
Dropped 4140 nodes (cum <= 36552)
Showing top 10 nodes out of 520
      flat  flat%   sum%        cum   cum%
     84262  1.15%  1.15%     100798  1.38%  github.com/cockroachdb/cockroach/pkg/util/tracing.(*Tracer).startSpanGeneric
     58556   0.8%  1.95%      58556   0.8%  context.WithValue
    -30415  0.42%  1.54%     -30410  0.42%  github.com/cockroachdb/cockroach/pkg/storage/enginepb.(*MVCCMetadata).Unmarshal
    -24898  0.34%  1.20%     -54566  0.75%  github.com/cockroachdb/cockroach/pkg/storage.(*MVCCValueMerger).Finish
    -20780  0.28%  0.91%     -20780  0.28%  github.com/cockroachdb/cockroach/pkg/roachpb.(*InternalTimeSeriesData).Unmarshal
     19986  0.27%  1.19%      19986  0.27%  github.com/gogo/protobuf/types.TimestampProto
     16538  0.23%  1.41%      16538  0.23%  github.com/cockroachdb/cockroach/pkg/util/tracing.(*crdbSpan).setTagLocked (inline)
    -15139  0.21%  1.21%     -66197  0.91%  github.com/cockroachdb/cockroach/pkg/storage.(*MVCCValueMerger).deserializeMVCCValueAndAppend
     13620  0.19%  1.39%      13620  0.19%  github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Meta
    -12258  0.17%  1.22%     -12258  0.17%  google.golang.org/grpc/metadata.Pairs
```

There's a little bit of small fry but the dominant term `startSpanGeneric` won't budge much; that is just the number of trace spans necessary for always-on tracing.

knz reviewed (commented):
I'm generally ok with the approach.

some requests:

- can you include the pprof alloc output when `trace.mode` is set to `background`, so that we have some high level intuition about the cost (maybe also some before/after measure for go test alloc/cpu   on some minimal server unit test with tracing enabled)
-  see below about the request to a doc writer
- is there a unit test for ensuring that the setting does, in fact, enable always-on tracing?

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-:-MOkgmL19RXVLUgHJG1-:b-2su9vv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/util/tracing/span.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgNuSEgUh-O8LzAKq:-MOkgNuT6EoiAwX9La-b:bjooebn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/span.go#L327)):*
> ```Go
> 	*toSort = result[1:]
> 	sort.Sort(toSort)
> 	sortPool.Put(toSort)
> ```

don't you need to truncate the slice before you put it back? to enable releasing the individual entries?
___
*[pkg/util/tracing/tracer.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgaly1u5GeYvlClQO:-MOkgaly1u5GeYvlClQP:bhxwu71) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/tracer.go#L69)):*
> ```Go
> 	`configures the CockroachDB-internal tracing subsystem.
> 
> If set to 'background', trace spans will be created for all operations, but
> ```

even though it is not _public_, I think it's worth copying this explanation to the commit message, then indicate in a release note to the doc writer we want this to be propagated to our internal troubleshooting guides (we have those now)


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> can you include the pprof alloc output when trace.mode is set to background, so that we have some high level intuition about the cost (maybe also some before/after measure for go test alloc/cpu on some minimal server unit test with tracing enabled)

I intentionally deferred this to more performance-minded PRs, but you can get a sneak peek here: https://github.com/cockroachdb/cockroach/pull/57292#issuecomment-743313435

I prefer to post a comparison when the initial performance work has happened, which is when they start meaning something. There are various such optimizations in the pipeline.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-:-MOkjI8vB6d7KMs8KaZU:b-pctbi3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/util/tracing/span.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgNuSEgUh-O8LzAKq:-MOkjkhv3kbeBXm-em2t:bp8yjw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/span.go#L327)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

don't you need to truncate the slice before you put it back? to enable releasing the individual entries?
</blockquote></details>

Yikes, thanks for catching that. Done.
___
*[pkg/util/tracing/tracer.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgaly1u5GeYvlClQO:-MOkjqU4AdE5FEqqW_3Q:b-sb1pf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/tracer.go#L69)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

even though it is not _public_, I think it's worth copying this explanation to the commit message, then indicate in a release note to the doc writer we want this to be propagated to our internal troubleshooting guides (we have those now)
</blockquote></details>

What would you want to have propagated? I am planning to flip the default before the new release and remove the setting. Besides, it's not like setting this to on will help troubleshooting at all, as the background spans never receive any information, nor can they be viewed anywhere.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> I prefer to post a comparison when the initial performance work has happened, which is when they start meaning something. There are various such optimizations in the pipeline.

ok cool that works

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-:-MOkm80t4kxIDsfMbgtq:b7zdn0a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/util/tracing/tracer.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgaly1u5GeYvlClQO:-MOklynBBGiVEj4QILvx:b6b37t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/tracer.go#L69)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

What would you want to have propagated? I am planning to flip the default before the new release and remove the setting. Besides, it's not like setting this to on will help troubleshooting at all, as the background spans never receive any information, nor can they be viewed anywhere.
</blockquote></details>

hm ok - but then I am confused about why you bothered to write two full paragraphs of explanations here if you believe they have no audience.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57292)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/util/tracing/tracer.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgaly1u5GeYvlClQO:-MOkz-0uEMM4Hz-7a5qZ:b-sw5vjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/tracer.go#L69)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

hm ok - but then I am confused about why you bothered to write two full paragraphs of explanations here if you believe they have no audience.
</blockquote></details>

Heh, are you telling me that I wrote too much documentation? ;-)

Just kidding, I see how that would be confusing. The reason is simple: when I first wrote this I wasn't quite sure where it was going yet and was also still explaining it to myself. And now that it's there, it does seem useful, even if we ultimately don't think it'll get too much airplay. I see it as eng-internal documentation while it exists.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-:-MOl-Ss71N_fuX_HDYnR:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/util/tracing/tracer.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57292#-MOkgaly1u5GeYvlClQO:-MOl-CjhAjS3PqCrT1IK:b-frrzzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6282b40b5c87dd7927d1be3e78a2dbfb5f0856dc/pkg/util/tracing/tracer.go#L69)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Heh, are you telling me that I wrote too much documentation? ;-)

Just kidding, I see how that would be confusing. The reason is simple: when I first wrote this I wasn't quite sure where it was going yet and was also still explaining it to myself. And now that it's there, it does seem useful, even if we ultimately don't think it'll get too much airplay. I see it as eng-internal documentation while it exists.
</blockquote></details>

point taken

meanwhile, I'd encourage you to appreciate how much our docs team has grown and how much willingness we have to start documenting internal processes / systems to help with support and escalations. It may not be relevant in this particular case, but we can+should involve doc writers more, even for things that are not user-facing.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=knz

TFTR!
